x/warp/contract.go

```diff
@@ -0,0 +1,278 @@
+	DestinationChainID [32]byte
+	DestinationAddress [32]byte
+	Payload            []byte
+}
+
+// PackGetBlockchainID packs the include selector (first 4 func signature bytes).
+// This function is mostly used for tests.
+func PackGetBlockchainID() ([]byte, error) {
+	return WarpABI.Pack("getBlockchainID")
+}
+
+// PackGetBlockchainIDOutput attempts to pack given blockchainID of type [32]byte
+// to conform the ABI outputs.
+func PackGetBlockchainIDOutput(blockchainID [32]byte) ([]byte, error) {
+	return WarpABI.PackOutput("getBlockchainID", blockchainID)
+}
+
+// getBlockchainID returns the snow Chain Context ChainID of this blockchain.
+func getBlockchainID(accessibleState contract.AccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	if remainingGas, err = contract.DeductGas(suppliedGas, GetBlockchainIDGasCost); err != nil {
+		return nil, 0, err
+	}
+	packedOutput, err := PackGetBlockchainIDOutput(accessibleState.GetSnowContext().ChainID)
+	if err != nil {
+		return nil, remainingGas, err
+	}
+
+	// Return the packed output and the remaining gas
+	return packedOutput, remainingGas, nil
+}
+
+// PackGetVerifiedWarpMessage packs the calldata for the getVerifiedWarpMessage function
+// This function is mostly used for tests.
+func PackGetVerifiedWarpMessage() ([]byte, error) {
+	return WarpABI.Pack("getVerifiedWarpMessage")
+}
+
+// PackGetVerifiedWarpMessageOutput attempts to pack given [outputStruct] of type GetVerifiedWarpMessageOutput
+// to conform the ABI outputs.
+func PackGetVerifiedWarpMessageOutput(outputStruct GetVerifiedWarpMessageOutput) ([]byte, error) {
+	return WarpABI.PackOutput("getVerifiedWarpMessage",
+		outputStruct.Message,
+		outputStruct.Exists,
+	)
+}
+
+// getVerifiedWarpMessage retrieves the pre-verified warp message from the predicate storage slots and returns
+// the expected ABI encoding of the message to the caller.
+func getVerifiedWarpMessage(accessibleState contract.AccessibleState, caller common.Address, addr common.Address, _ []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	remainingGas = suppliedGas
+	// XXX Note: there is no base cost for retrieving a verified warp message. Instead, we charge for each piece of gas,
+	// prior to each execution step.
+	// Ignore input since there are no arguments
+	predicateBytes, exists := accessibleState.GetStateDB().GetPredicateStorageSlots(ContractAddress)
+	// If there is no such value, return false to the caller.
+	if !exists {
+		packedOutput, err := PackGetVerifiedWarpMessageOutput(GetVerifiedWarpMessageOutput{
+			Exists: false,
+		})
+		if err != nil {
+			return nil, remainingGas, err
+		}
+		return packedOutput, remainingGas, nil
+	}
+
+	msgBytesGas, overflow := math.SafeMul(GasCostPerWarpMessageBytes, uint64(len(predicateBytes)))
+	if overflow {
+		return nil, remainingGas, vmerrs.ErrOutOfGas
+	}
```

Comment:
It seems we charged for this gas once in PredicateGas? (i.e, actual cost per byte is 2* GasCostPerWarpMessageBytes)

Reply:
Could we add a comment about this?

Reply:
Adding